tests/Client.hs

```diff
@@ -0,0 +1,232 @@
+    encodePublicKeyCredentialRequestOptions,
+    encodeRequestedPublicKeyCredential,
+  )
+import qualified Crypto.Fido2.Model.JavaScript.Types as JS
+import Crypto.Fido2.Operations.Attestation (allSupportedFormats)
+import qualified Crypto.Fido2.Operations.Attestation as Fido2
+import qualified Crypto.Fido2.PublicKey as PublicKey
+import Crypto.Hash (hash, hashlazy)
+import Crypto.Random (MonadRandom)
+import Data.Aeson (encode)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base64.URL as Base64
+import Data.Either (fromRight, isRight)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import qualified Data.Set as Set
+import Data.Text.Encoding (decodeUtf8)
+import Data.Validation (toEither)
+import System.Random.Stateful (globalStdGen, uniformM)
+import Test.Hspec (SpecWith, describe, it, shouldSatisfy)
+
+-- | Emulates the client-side operation for attestation given an authenticator.
+-- MonadRandom is required during the geneation of the new credentials.
+clientAttestation :: (MonadRandom m, MonadFail m) => JS.PublicKeyCredentialCreationOptions -> Authenticator -> m (JS.CreatedPublicKeyCredential, Authenticator)
+clientAttestation options authenticator = do
+  let M.PublicKeyCredentialCreationOptions {..} =
+        fromRight (error "Test: could not decode creation options") $ decodePublicKeyCredentialCreationOptions options
+      clientDataBS =
+        encode
+          JS.ClientDataJSON
+            { JS.typ = "webauthn.create",
+              JS.challenge = decodeUtf8 . Base64.encode $ M.unChallenge pkcocChallenge,
+              JS.origin = "https://localhost:8080",
+              JS.crossOrigin = Nothing
+            }
+      clientDataHash = M.ClientDataHash $ hashlazy clientDataBS
+  (attestationObject, authenticator') <-
+    authenticatorMakeCredential
+      authenticator
+      clientDataHash
+      pkcocRp
+      pkcocUser
+      True
+      True
+      True
+      pkcocPubKeyCredParams
+      pkcocExcludeCredentials
+      False
+      pkcocExtensions
+  let response =
+        encodeCreatedPublicKeyCredential
+          M.PublicKeyCredential
+            { M.pkcIdentifier = M.acdCredentialId . M.adAttestedCredentialData $ M.aoAuthData attestationObject,
+              M.pkcResponse =
+                M.AuthenticatorAttestationResponse
+                  { M.arcClientData =
+                      M.CollectedClientData
+                        { M.ccdChallenge = pkcocChallenge,
+                          M.ccdOrigin = M.Origin "https://localhost:8080",
+                          M.ccdCrossOrigin = Just True,
+                          M.ccdHash = M.ClientDataHash $ hashlazy clientDataBS
+                        },
+                    M.arcAttestationObject = attestationObject,
+                    M.arcTransports = Set.fromList [M.AuthenticatorTransportUSB, M.AuthenticatorTransportBLE, M.AuthenticatorTransportNFC, M.AuthenticatorTransportInternal]
```

Comment:
How come these values?

Reply:
Because our test authenticator is believed to support all transports.